generate/authorize.js

```diff
@@ -0,0 +1,834 @@
+  // create empty userRoles and docRoles objects
+  // as default values, which are used
+  // if there is not @authorize directive
+  const userRoles = {};
+  const docRoles =  {};
+  let roleFieldNamesFound = [];
+  CODE_MODES.forEach(mode => userRoles[mode] = []);
+  CODE_MODES.forEach(mode => docRoles[mode] = []);
+
+  // check if there is an @authorize directive
+  if ( authorize ) {
+    // then re-determine the userRoles and docRoles 
+    // from the @authorize tag of the type definition
+    const allRolesArguments = inputSchema.definitions[0].directives[0].arguments || {};
+    const allRoles = getAllRoles ( allRolesArguments, inputSchema );
+    allRoles.forEach(role => {
+      switch (role.type) {
+
+        case USER_ROLE:
+          // check, if there is already another userRole field
+          if (roleFieldNamesFound.length > 0 && 
+              role.roleFieldName !== '' &&
+              !roleFieldNamesFound.includes(role.roleFieldName)) {
+
+            // We allow only one field which keeps all userRoles
+            throw new Error(`Please adjust type definition, that there is only ONE field, 
+              which keeps all user roles. You've tried to add a second userRole field: '${role.roleFieldName}',
+              but there is already another userRole field: '${roleFieldNamesFound[0]}' defined.
+              Please try instead: '${roleFieldNamesFound[0]}: String @authRole(for: ["otherRole", "${role.roleName}"])'`);
+          }
+          if (role.roleFieldName !== '') roleFieldNamesFound.push(role.roleFieldName);
+
+          Object.keys(role.modes).forEach(mode => {
+            if(role.modes[mode]) userRoles[mode].push(role.roleName);
+          });
+          break;
+
+        case DOC_ROLE:
+          Object.keys(role.modes).forEach(mode => {
+            if(role.modes[mode]) docRoles[mode].push(role.roleName);
+          });
+          break;
+      }
+    });
+  }
+
+  return {
+    userRoles,
+    docRoles,
+    roleFieldName: roleFieldNamesFound.length > 0 ? roleFieldNamesFound[0] : '',
+  };
+
+}
+
+/*
+ * get the roles from the @authorize directive
+ * by reading the input schema's abstract syntax tree
+ * to get the roles and their authorized modes
+ * @param {object} allRolesArguments
+ * @return {array} allRoles
+ */
+function getAllRoles ( allRolesArguments = [], inputSchema ) {
+  /*  Example:
+      @authorize(
+      //role:  roleModes: [mode.value]   
+        admin: ["create", "read", "update", "delete"]
+        this: ["read", "update", "delete"]
+      ) 
+  */
+  let allRoles = [];
+
+  // get all Roles of the type's @authorize directives
+  // e.g. 'admin', 'this'
+  allRolesArguments.forEach(roleArgument => {
+    // new role found
+    const role = {};
+
+    // check if it is a valid role
+    if (roleArgument.kind === ARGUMENT &&
+        roleArgument.name && 
+        roleArgument.name.kind === NAME &&
+        roleArgument.name.value &&
+        roleArgument.name.value !== ''){
```

Comment:
In general do we need to be this defensive with this AST reading code? Aren't there certain guarantees about what comes out of `graphql`?

Reply:
It makes this code quite hard to read having all these null checks.

Reply:
changed it